flasksrc/session.py

```diff
@@ -607,6 +625,105 @@ def put(self, toynet_session_id, **kwargs):
             }, 200
 
 
+class ToyNetSessionByIdCreateRouterInterface(MethodResource):
+    @use_kwargs(ToyNetSessionByIdCreateRouterInterfacePutReq)
+    @marshal_with(ToyNetSessionByIdCreateRouterInterfacePutResp)
+    def put(self, toynet_session_id, **kwargs):
+        try:
+            req = ToyNetSessionByIdCreateRouterInterfacePutReq().load(kwargs)
+        except ValidationError as e:
+            abort(400, message='Invalid request: {}'.format(e))
+
+        # Separate validation from Marshmallow
+        if 'ip' not in req:
+            abort(400, message='Missing ip from req')
+        elif 'name' not in req:
+            abort(400, message='Missing name from req')
+
+        orig_topology = getTopologyFromDb(toynet_session_id)['topology']
+        root = ET.fromstring(orig_topology)
+
+        # Get the router to add the interface to, note that we do not validate
+        # that the IP address is "good" (eg: not a duplicate of another
+        # interface)
+        router = None
+        router_list = root.find('routerList').findall('router')
+
+        for device in router_list:
+            if device.get('name') == req['name']:
+                router = device
+
+        if router is None:
+            abort(400, message=f'Router {req["name"]} does not exist')
+
+        # Create XML elements and update topology
+        interface = ET.Element('intf')
+        interface.text = req['ip']
+        router.append(interface)
+        new_topology = ET.tostring(root).decode('utf-8')
+
+        # Send to mininet and update DB, these functions abort the REST call on
+        # failure
+        sendTopoToMininet(toynet_session_id, new_topology)
+        updateTopoInDb(toynet_session_id, new_topology)
+
+        return {
+            }, 200
+
+
+class ToyNetSessionByIdDeleteRouterInterface(MethodResource):
+    @use_kwargs(ToyNetSessionByIdDeleteRouterInterfacePutReq)
+    @marshal_with(ToyNetSessionByIdDeleteRouterInterfacePutResp)
+    def put(self, toynet_session_id, **kwargs):
+        try:
+            req = ToyNetSessionByIdDeleteRouterInterfacePutReq().load(kwargs)
+        except ValidationError as e:
+            abort(400, message='Invalid request: {}'.format(e))
+
+        # Separate validation from Marshmallow
+        if 'ip' not in req:
+            abort(400, message='Missing ip from req')
+        elif 'name' not in req:
+            abort(400, message='Missing name from req')
+
+        orig_topology = getTopologyFromDb(toynet_session_id)['topology']
```

Comment:
Let's `try-catch` the exception in this function & `abort`?

Reply:
I prefer this method because it has the error control logic for all inputs in one place and we don't need to have either: many try-excepts on each attempted variable access, or inconsistent use of try-catch by only checking the first access. If we were to change it we should do the same for the other methods. Thoughts?

Reply:
I think your logic checks out, thanks :)